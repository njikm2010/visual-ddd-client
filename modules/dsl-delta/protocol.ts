/**
 * 协议设计
 */
import type { BusinessDomainDSL, NameDSL } from '@/modules/domain/api/dsl/interface';
import type { ScenarioDSL } from '@/modules/scenario/api/dsl/interface';

export type * from '@/modules/domain/api/dsl/interface';
export type * from '@/modules/scenario/api/dsl/interface';

export interface WithVersion {
  /**
   * 版本号
   */
  version: string;
}

/**
 * 业务域
 */
export interface NormalizedBusinessDomainDSL extends NameDSL, WithVersion, BusinessDomainDSL {}

/**
 * 业务场景
 */
export interface NormalizedScenarioDSL extends NameDSL, WithVersion, ScenarioDSL {}

/**
 * 完整的应用 DSL
 */
export interface ApplicationDSL extends NameDSL, WithVersion {
  /**
   * 业务域
   */
  businessDomains: NormalizedBusinessDomainDSL[];

  /**
   * 业务场景
   */
  businessScenarios: NormalizedScenarioDSL[];
}

/**
 * 对象 diff 协议
 */
export namespace Delta {
  /**
   * 对象标记类型
   */
  export enum OP {
    /**
     * 无任何操作
     */
    OP_NONE = 0,

    /**
     * 标记节点及其子节点为新增
     */
    OP_NEW = 1,

    /**
     * 标记移除
     */
    OP_DELETE = 1 << 1,

    /**
     * 标记变动。
     */
    OP_UPDATE = 1 << 2,

    // 扩展标记
    /**
     * 对于位移的对象，比如从一个容器移动到另一个容器，会被标记为 Move
     * 目前仅在业务域的领域建模中可能存在这种情况
     */
    OP_MOVE = 1 << 3,
  }

  export interface PropertyUpdate {
    type: OP.OP_UPDATE;
    /**
     * 变更的属性
     */
    key: string;

    /**
     * 旧的值, 只有原始值会携带 oldValue
     * 如果是引用类型，这里为空
     */
    oldValue?: any;
  }

  export interface PropertyNew {
    type: OP.OP_NEW;
    key: string;
  }

  /**
   * 属性删除
   */
  export interface PropertyDelete {
    type: OP.OP_DELETE;
    key: string;

    /**
     * 被删除的值或对象
     */
    value: any;
  }

  export type PropertyDelta = PropertyUpdate | PropertyNew | PropertyDelete;

  /**
   * 对象变更标记
   */
  export interface Delta {
    /**
     * OP 位
     */
    op: number;

    /**
     * 属性 Delta
     */
    deltas: PropertyDelta[];
  }
}

// 注入 delta 标记到 NameDSL
declare module '@/modules/domain/api/dsl/interface' {
  export interface NameDSL {
    /**
     * Delta 标记
     */
    __delta?: Delta.Delta;
  }
}

/**
 * 值类型
 */
export enum ValueType {
  /**
   * 原子类型
   */
  Atom,

  /**
   * 不处理
   */
  Never,
}

/**
 * 对象结构定义
 */
export type ObjectMeta<T> = T extends {}
  ? {
      [K in keyof T]: ValueType | ObjectMeta<T[K]> | [ObjectMeta<T[K]>];
    }
  : ValueType;
