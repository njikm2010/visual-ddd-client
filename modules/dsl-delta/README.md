# DSL 比对算法以及标记形式

- 我们会将比对结果标记在 DSL 上，代码生成器根据 DSL + 标记 增量生成最终的代码。
- 标记只是代码生成器的'辅助信息'。可以用于指示代码的变更操作，比如**移除**和**重命名**。因为单纯从一个 DSL 副本中无法得出这些信息
- 为了更好的容错性，生成器仍然需要完整地遍历 DSL，而不是仅仅根据标记进行操作。因为代码和 DSL 之间可能会有不一致。

## 基本概念

- 原子值：即没有下级对象或者关联关系的值，如字符串、数字、布尔值、字符串数组等, 这些值可以直接进行值比较
- 对象：对象是主要的比对单位, 每个对象都有唯一的 id(uuid)

## 标记方式

```ts
{
  // 对象属性
  __delta: Delta;
}
```

扩展一个 `__delta` 属性，用于标记对象的变更情况

## Delta 对象:

```ts
{
  // new 该对象被标记为新增对象
  // delete 该对象被标记为删除
  // move 每个对象都有唯一的 id, 该对象的上下级关系发生了变更, 生成器可以从旧的 DSL 中查找就得上下文，然后进行移动操作
  // dirty 该对象已经脏，即本身或者子节点发生了变更, 对于需要性能优化的场景，可以通过这个属性来判断是否需要继续往下遍历
  // 这里可能是位操作，不排除后续还有其他状态的接入，或者并行的状态
  op: 'new' | 'delete' | 'move' | 'dirty' | 'update';

  // 当前组件的一些变更操作
  deltas: Array<
    | {
        type: 'update';
        key: string;
        oldValue: any;
      }
    | {
        type: 'new';
        key: string;
      }
    | {
        type: 'delete';
        key: string;
        value: 2;
      }
    | {
        type: 'dirty';
        key: string;
      }
  >;
}
```

对比两个对象：

如果没有 key，则是标记自身

| A                 | B                 | delta                                 | 对象标记  | 备注                                                                                     |
| ----------------- | ----------------- | ------------------------------------- | --------- | ---------------------------------------------------------------------------------------- |
| {a: 1}            | undefined         |                                       | new       | 全新的对象                                                                               |
| undefined         | {a: 1}            |                                       | delete    | 对象被删除                                                                               |
| {a: 1}            | {a: 2}            | {type: 'update', key: a, oldValue: 2} | dirty     | 属性变更                                                                                 |
| {a: 1, b: 2}      | {a: 1}            | {type: 'new', key: b }                | dirty     | 属性新增                                                                                 |
| {a: 1 }           | {a: 1, b: 2}      | {type: 'delete', key: b, value: 2}    | dirty     | 属性删除                                                                                 |
| {a: {a: 1}}       | {a: {b: 1}}       | {type: 'dirty', key: a },             | dirty     | 引用属性变更                                                                             |
| A -> {a: 1, b: 2} | B -> {a: 1, b: 3} |                                       | move, new | 对象移动, 作为一个全新的对象处理，生成器可以从其他地方移动代码文件，在此的基础上进行生成 |

---

对比两个引用数组:

- 因为每个对象都有唯一 id, 我们可以简单地计算出新增和删除
- 顺序会按照新的 DSL 进行排序
- 被移除的对象会放在队列尾部
- 对于新增的对象，可能会在堂兄弟之间查找，识别出移动操作

举个例子:

```
New: [A, B, C]
Old: [B, C, D]

NewBrother: [E, F, G]
OldBrother: [A, E, F]
```

```
New
[
	A with delta {op: move},
	B,
	C,
	D with delta {op: delete},
]
```

```
NewBrother
[
	E,
	F,
	G with delta {op: new},
	A with delta {op: delete},
]
```
