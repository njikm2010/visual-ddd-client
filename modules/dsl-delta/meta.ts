import { ObjectMeta, ValueType, NameDSL, ApplicationDSL, WithVersion } from './protocol';

export const NAME_DSL_META: ObjectMeta<NameDSL> = {
  uuid: ValueType.Atom,
  name: ValueType.Atom,
  title: ValueType.Atom,
  description: ValueType.Atom,
  meta: ValueType.Atom,
  __delta: ValueType.Never,
};

export const VERSION_DSL_META: ObjectMeta<WithVersion> = {
  version: ValueType.Atom,
};

/**
 * 应用 DSL meta
 */
export const APPLICATION_DSL_META: ObjectMeta<ApplicationDSL> = {
  ...NAME_DSL_META,
  ...VERSION_DSL_META,
  businessDomains: ValueType.Atom,
  businessScenarios: ValueType.Atom,
};
