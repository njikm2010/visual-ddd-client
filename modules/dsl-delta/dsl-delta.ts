import { objectDiff, markOP, isOPMarked } from './diff';
import { ApplicationDSL, Delta, DomainModelDSL, EntityDSL, NameDSL, ObjectMeta } from './protocol';
import { APPLICATION_DSL_META } from './meta';
import cloneDeep from 'lodash/cloneDeep';
import { NoopArray } from '@wakeapp/utils';

export function moveDetectForDomainModel(domain: DomainModelDSL) {
  const { aggregates } = domain;

  // 对象可能在聚合之间移动
  const objectsMarkDeleted: Map<string, NameDSL[]> = new Map();
  const objectsMarkNew: NameDSL[] = [];

  const markDeleted = (obj: NameDSL) => {
    const list = objectsMarkDeleted.get(obj.uuid);

    if (list) {
      list.push(obj);
    } else {
      objectsMarkDeleted.set(obj.uuid, [obj]);
    }
  };

  const checkList = (list: NameDSL[]) => {
    for (const i of list) {
      if (isOPMarked(i.__delta?.op, Delta.OP.OP_NEW)) {
        objectsMarkNew.push(i);
      } else if (isOPMarked(i.__delta?.op, Delta.OP.OP_DELETE)) {
        // 标记删除
        markDeleted(i);
      }
    }
  };

  for (const agg of aggregates) {
    // 聚合根
    if (isOPMarked(agg.root.__delta?.op, Delta.OP.OP_NEW)) {
      objectsMarkNew.push(agg.root);
    }

    const rootDelete = agg.__delta?.deltas.find(i => i.type === Delta.OP.OP_DELETE && i.key === 'root');
    if (rootDelete) {
      markDeleted((rootDelete as Delta.PropertyDelete).value as EntityDSL);
    }

    checkList(agg.entities || NoopArray);
    checkList(agg.valueObjects || NoopArray);
    checkList(agg.enums || NoopArray);
    checkList(agg.commands || NoopArray);
  }

  for (const obj of objectsMarkNew) {
    if (objectsMarkDeleted.has(obj.uuid)) {
      // 监听到移动
      obj.__delta!.op = markOP(obj.__delta!.op, Delta.OP.OP_MOVE);

      const deleted = objectsMarkDeleted.get(obj.uuid);
      deleted?.forEach(i => {
        i.__delta!.op = markOP(i.__delta!.op, Delta.OP.OP_MOVE);
      });
    }
  }
}

/**
 * 对象移动检测
 * @param dsl
 */
export function moveDetected(dsl: ApplicationDSL) {
  dsl.businessDomains.forEach(domain => {
    moveDetectForDomainModel(domain.domainModel);
  });
}

/**
 * 生成 DSL delta
 */
export function dslDelta(income: ApplicationDSL, old: ApplicationDSL) {
  const source = cloneDeep(income);
  const oldSource = cloneDeep(old);

  objectDiff(source, oldSource, APPLICATION_DSL_META as ObjectMeta<any>);

  moveDetected(source);

  return source;
}
