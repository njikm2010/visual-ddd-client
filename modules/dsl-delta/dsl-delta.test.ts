import { dslDelta, moveDetectForDomainModel } from './dsl-delta';
import { ApplicationDSL, Delta, DomainModelDSL, NormalizedBusinessDomainDSL } from './protocol';

describe('move detect', () => {
  test('root move to entities', () => {
    const domain: DomainModelDSL = {
      aggregates: [
        {
          uuid: '1',
          name: '1',
          root: {
            uuid: '01',
            name: 'User',
            meta: {},
            id: {
              uuid: '011',
              name: 'id',
              description: '',
              meta: {},
              access: 'public',
              type: 'Long',
            },
            properties: [],
            methods: [],
            __delta: {
              op: Delta.OP.OP_NEW,
              deltas: [],
            },
          },
          entities: [
            {
              uuid: '02',
              name: 'Bar',
              meta: {},
              id: {
                uuid: '021',
                name: 'id',
                description: '',
                meta: {},
                access: 'public',
                type: 'Long',
              },
              properties: [],
              methods: [],
              __delta: {
                op: Delta.OP.OP_NEW,
                deltas: [],
              },
            },
          ],
          commands: [],
          __delta: {
            op: Delta.OP.OP_UPDATE,
            deltas: [
              {
                type: Delta.OP.OP_DELETE,
                key: 'root',
                value: {
                  uuid: '02',
                  name: 'Bar',
                  meta: {},
                  id: {
                    uuid: '021',
                    name: 'id',
                    description: '',
                    meta: {},
                    access: 'public',
                    type: 'Long',
                  },
                  properties: [],
                  methods: [],
                  __delta: {
                    op: Delta.OP.OP_DELETE,
                    deltas: [],
                  },
                },
              },
            ],
          },
        },
      ],
    };

    moveDetectForDomainModel(domain);

    // 标记为 new
    expect(domain.aggregates[0].entities?.[0]?.__delta?.op).toBe(Delta.OP.OP_MOVE | Delta.OP.OP_NEW);
    expect(domain.aggregates[0].root.__delta?.op).toBe(Delta.OP.OP_NEW);
    expect((domain.aggregates[0].__delta?.deltas[0] as Delta.PropertyDelete).value.__delta.op).toBe(
      Delta.OP.OP_DELETE | Delta.OP.OP_MOVE
    );
  });

  test('entities move to root', () => {
    const domain: DomainModelDSL = {
      aggregates: [
        {
          uuid: '1',
          name: '1',
          root: {
            uuid: '01',
            name: 'User',
            meta: {},
            id: {
              uuid: '011',
              name: 'id',
              description: '',
              meta: {},
              access: 'public',
              type: 'Long',
            },
            properties: [],
            methods: [],
            __delta: {
              op: Delta.OP.OP_NEW,
              deltas: [],
            },
          },
          entities: [
            {
              uuid: '02',
              name: 'Bar',
              meta: {},
              id: {
                uuid: '021',
                name: 'id',
                description: '',
                meta: {},
                access: 'public',
                type: 'Long',
              },
              properties: [],
              methods: [],
              __delta: {
                op: Delta.OP.OP_NEW,
                deltas: [],
              },
            },
            {
              uuid: '01',
              name: 'User',
              meta: {},
              id: {
                uuid: '011',
                name: 'id',
                description: '',
                meta: {},
                access: 'public',
                type: 'Long',
              },
              properties: [],
              methods: [],
              __delta: {
                op: Delta.OP.OP_DELETE,
                deltas: [],
              },
            },
          ],
          commands: [],
          __delta: {
            op: Delta.OP.OP_UPDATE,
            deltas: [
              {
                type: Delta.OP.OP_DELETE,
                key: 'root',
                value: {
                  uuid: '03',
                  name: 'Baz',
                  meta: {},
                  id: {
                    uuid: '031',
                    name: 'id',
                    description: '',
                    meta: {},
                    access: 'public',
                    type: 'Long',
                  },
                  properties: [],
                  methods: [],
                  __delta: {
                    op: Delta.OP.OP_DELETE,
                    deltas: [],
                  },
                },
              },
            ],
          },
        },
      ],
    };

    moveDetectForDomainModel(domain);
    expect(domain.aggregates[0].root.__delta?.op).toBe(Delta.OP.OP_NEW | Delta.OP.OP_MOVE);
    expect(domain.aggregates[0].entities?.[1].__delta?.op).toBe(Delta.OP.OP_DELETE | Delta.OP.OP_MOVE);
  });

  test('move between aggregation', () => {
    const domain: DomainModelDSL = {
      aggregates: [
        {
          uuid: '1',
          name: '1',
          root: {} as any,
          entities: [
            {
              uuid: '02',
              name: 'Bar',
              meta: {},
              id: {
                uuid: '021',
                name: 'id',
                description: '',
                meta: {},
                access: 'public',
                type: 'Long',
              },
              properties: [],
              methods: [],
              __delta: {
                op: Delta.OP.OP_NEW,
                deltas: [],
              },
            },
          ],
          commands: [],
        },
        {
          uuid: '2',
          name: '2',
          root: {} as any,
          entities: [
            {
              uuid: '02',
              name: 'Bar',
              meta: {},
              id: {
                uuid: '021',
                name: 'id',
                description: '',
                meta: {},
                access: 'public',
                type: 'Long',
              },
              properties: [],
              methods: [],
              __delta: {
                op: Delta.OP.OP_DELETE,
                deltas: [],
              },
            },
          ],
          commands: [],
        },
      ],
    };

    moveDetectForDomainModel(domain);

    expect(domain.aggregates[0].entities?.[0]?.__delta?.op).toBe(Delta.OP.OP_MOVE | Delta.OP.OP_NEW);
    expect(domain.aggregates[1].entities?.[0]?.__delta?.op).toBe(Delta.OP.OP_DELETE | Delta.OP.OP_MOVE);
  });
});

test('simple', () => {
  const sourceDomain: NormalizedBusinessDomainDSL = {
    uuid: 'simple',
    name: 'simple',
    version: '0.1.0',
    domainModel: {
      aggregates: [
        {
          uuid: '51d6a0f1-cfef-4723-9aae-2bed0ecf0e9d',
          title: '用户',
          name: 'User',
          meta: {},
          root: {
            uuid: '6c75e505-03d9-40aa-9961-bc5e37473da1',
            title: '用户',
            name: 'User',
            meta: {},
            id: {
              uuid: 'd5ac8416-5fe0-4407-b7d9-c22f207ab3bb',
              title: '用户 ID',
              name: 'id',
              description: '',
              meta: {},
              access: 'public',
              type: 'Long',
            },
            properties: [
              {
                uuid: '9ee1c90d-ac17-45be-897a-5757d1b5f9d7',
                title: '名称',
                name: 'name',
                description: '',
                meta: {},
                access: 'public',
                type: 'String',
              },
            ],
            methods: [],
          },
          entities: [],
          valueObjects: [],
          enums: [],
          commands: [
            {
              uuid: 'c89fb0b5-7604-4863-a754-af5a36d6e80f',
              title: '新增用户',
              name: 'UserCreate',
              description: '',
              meta: {},
              rules: [],
              category: 'create',
              repository: 'create',
              eventSendable: false,
              source: [{ type: 'http' }, { type: 'rpc' }],
              properties: [
                {
                  uuid: '1a925506-00f7-4aef-bcce-836215911c00',
                  title: '名称',
                  name: 'name',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'String',
                },
              ],
              event: {
                uuid: 'c89fb0b5-7604-4863-a754-af5a36d6e80f',
                name: 'UserCreateEvent',
                properties: [
                  {
                    uuid: '1537b249-0aba-4177-8b76-93b1b4a7b535',
                    title: '用户 ID',
                    name: 'id',
                    description: '',
                    meta: {},
                    access: 'public',
                    type: 'Long',
                  },
                ],
              },
              return: { type: 'void' },
            },
            {
              uuid: 'e07d602b-b97a-492c-a46e-8798ff3bf8d0',
              title: '编辑用户',
              name: 'UserModify',
              description: '',
              meta: {},
              rules: [],
              category: 'modify',
              repository: 'modify',
              eventSendable: false,
              source: [{ type: 'http' }, { type: 'rpc' }],
              properties: [
                {
                  uuid: '30c6d04a-8ea0-466f-ab4e-d55d9e59d5d7',
                  title: '用户 ID',
                  name: 'id',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'Long',
                },
                {
                  uuid: '39881d97-37e2-4f33-9220-24827e585aee',
                  title: '名称',
                  name: 'name',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'String',
                },
              ],
              event: { uuid: 'e07d602b-b97a-492c-a46e-8798ff3bf8d0', name: 'UserModifyEvent', properties: [] },
              return: { type: 'void' },
            },
            {
              uuid: 'a69ea6f4-cc38-4597-b47e-51bf8081f5ea',
              title: '删除用户',
              name: 'UserRemove',
              description: '',
              meta: {},
              rules: [],
              category: 'remove',
              repository: 'remove',
              eventSendable: false,
              source: [{ type: 'http' }, { type: 'rpc' }],
              properties: [
                {
                  uuid: '25106aaf-6596-44bd-81f5-053a33f5d440',
                  title: '用户 ID',
                  name: 'id',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'Long',
                },
              ],
              event: { uuid: 'a69ea6f4-cc38-4597-b47e-51bf8081f5ea', name: 'UserRemoveEvent', properties: [] },
              return: { type: 'void' },
            },
          ],
        },
      ],
    },
    queryModel: { queries: [], dtos: [] },
    dataModel: { dataObjects: [], references: [] },
    objectMapper: { mappers: [] },
    vision: '',
    ubiquitousLanguage: [
      {
        conception: '资源',
        englishName: 'Resource',
        definition: 'BD进行销售活动的对象',
        example: '',
        restraint: '',
        uuid: 'bf0b0ce7-6028-41d8-89d6-d096485f3f7d',
      },
      {
        conception: '公海',
        englishName: 'PublicSea',
        definition: '业务目标限定的模拟资源全集',
        example: '业务目标是团购买单合作，限定POI为到餐全国和综合58同城的',
        restraint: '与业务目标对应，比如交易公海是为了做交易，推广公海是为了推广',
        uuid: 'c9bae1d0-d028-418e-98f6-6a3990eec09a',
      },
    ],
  };
  const applicationSource: ApplicationDSL = {
    uuid: 'application',
    version: '1.0.0',
    name: '应用',
    businessDomains: [sourceDomain],
    businessScenarios: [],
  };

  // 没有任何变更
  expect(dslDelta(applicationSource, applicationSource)).toEqual(applicationSource);

  const newDomain: NormalizedBusinessDomainDSL = {
    uuid: 'simple',
    name: 'simple',
    version: '0.1.0',
    domainModel: {
      aggregates: [
        {
          uuid: '51d6a0f1-cfef-4723-9aae-2bed0ecf0e9d',
          title: '用户',
          name: 'User',
          meta: {},
          root: {
            uuid: '6c75e505-03d9-40aa-9961-bc5e37473da1',
            title: '用户',
            name: 'User',
            meta: {},
            id: {
              uuid: 'd5ac8416-5fe0-4407-b7d9-c22f207ab3bb',
              title: '用户 ID',
              name: 'id',
              description: '',
              meta: {},
              access: 'public',
              type: 'Long',
            },
            properties: [
              {
                uuid: '9ee1c90d-ac17-45be-897a-5757d1b5f9d7',
                title: '名称',
                name: 'name',
                description: '',
                meta: {},
                access: 'public',
                type: 'String',
              },
              {
                uuid: '5741d468-0056-4850-a11c-ba9d9c03b8e6',
                title: '未命名',
                name: 'age',
                description: '',
                meta: {},
                access: 'public',
                type: 'Integer',
              },
            ],
            methods: [],
          },
          entities: [
            {
              uuid: 'copy-407908de-4cdc-4076-9ec7-958295d9b3e8',
              title: '订单',
              name: 'Order',
              meta: {},
              id: {
                uuid: 'd5ac8416-5fe0-4407-b7d9-c22f207ab3bb',
                title: '用户 ID',
                name: 'id',
                description: '',
                meta: {},
                access: 'public',
                type: 'Long',
              },
              properties: [
                {
                  uuid: 'a59edff9-b80b-43a2-b65d-f1a359acee6c',
                  title: '订单名称',
                  name: 'orderId',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'String',
                },
                {
                  uuid: '6159b563-a6f2-478f-864a-828bca126793',
                  title: '关联用户',
                  name: 'userId',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'Long',
                },
              ],
              methods: [],
            },
          ],
          valueObjects: [],
          enums: [],
          commands: [
            {
              uuid: 'c89fb0b5-7604-4863-a754-af5a36d6e80f',
              title: '新增用户',
              name: 'UserCreate',
              description: '',
              meta: {},
              rules: [],
              category: 'create',
              repository: 'create',
              eventSendable: false,
              source: [{ type: 'http' }, { type: 'rpc' }],
              properties: [
                {
                  uuid: '1a925506-00f7-4aef-bcce-836215911c00',
                  title: '名称',
                  name: 'name',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'String',
                },
              ],
              event: {
                uuid: 'c89fb0b5-7604-4863-a754-af5a36d6e80f',
                name: 'UserCreateEvent',
                properties: [
                  {
                    uuid: '1537b249-0aba-4177-8b76-93b1b4a7b535',
                    title: '用户 ID',
                    name: 'id',
                    description: '',
                    meta: {},
                    access: 'public',
                    type: 'Long',
                  },
                ],
              },
              return: { type: 'void' },
            },
            {
              uuid: 'e07d602b-b97a-492c-a46e-8798ff3bf8d0',
              title: '编辑用户',
              name: 'UserModify',
              description: '',
              meta: {},
              rules: [],
              category: 'modify',
              repository: 'modify',
              eventSendable: false,
              source: [{ type: 'http' }, { type: 'rpc' }],
              properties: [
                {
                  uuid: '30c6d04a-8ea0-466f-ab4e-d55d9e59d5d7',
                  title: '用户 ID',
                  name: 'id',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'Long',
                },
                {
                  uuid: '39881d97-37e2-4f33-9220-24827e585aee',
                  title: '名称',
                  name: 'name',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'String',
                },
              ],
              event: { uuid: 'e07d602b-b97a-492c-a46e-8798ff3bf8d0', name: 'UserModifyEvent', properties: [] },
              return: { type: 'void' },
            },
            {
              uuid: 'a69ea6f4-cc38-4597-b47e-51bf8081f5ea',
              title: '删除用户',
              name: 'UserRemove',
              description: '',
              meta: {},
              rules: [],
              category: 'remove',
              repository: 'remove',
              eventSendable: false,
              source: [{ type: 'http' }, { type: 'rpc' }],
              properties: [
                {
                  uuid: '25106aaf-6596-44bd-81f5-053a33f5d440',
                  title: '用户 ID',
                  name: 'id',
                  description: '',
                  meta: {},
                  access: 'public',
                  type: 'Long',
                },
              ],
              event: { uuid: 'a69ea6f4-cc38-4597-b47e-51bf8081f5ea', name: 'UserRemoveEvent', properties: [] },
              return: { type: 'void' },
            },
          ],
        },
      ],
    },
    queryModel: { queries: [], dtos: [] },
    dataModel: { dataObjects: [], references: [] },
    objectMapper: { mappers: [] },
    vision: '',
    ubiquitousLanguage: [
      {
        conception: '资源',
        englishName: 'Resource',
        definition: 'BD进行销售活动的对象',
        example: '',
        restraint: '',
        uuid: 'bf0b0ce7-6028-41d8-89d6-d096485f3f7d',
      },
      {
        conception: '公海',
        englishName: 'PublicSea',
        definition: '业务目标限定的模拟资源全集',
        example: '业务目标是团购买单合作，限定POI为到餐全国和综合58同城的',
        restraint: '与业务目标对应，比如交易公海是为了做交易，推广公海是为了推广',
        uuid: 'c9bae1d0-d028-418e-98f6-6a3990eec09a',
      },
    ],
  };

  const newApplicationSource: ApplicationDSL = {
    uuid: 'application',
    version: '1.0.0',
    name: '应用',
    businessDomains: [newDomain],
    businessScenarios: [],
  };

  expect(dslDelta(newApplicationSource, applicationSource)).toMatchSnapshot();
});
