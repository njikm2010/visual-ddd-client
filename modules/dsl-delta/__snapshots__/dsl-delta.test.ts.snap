// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple 1`] = `
{
  "businessDomains": [
    {
      "__delta": {
        "deltas": [
          {
            "key": "domainModel",
            "type": 4,
          },
        ],
        "op": 4,
      },
      "dataModel": {
        "dataObjects": [],
        "references": [],
      },
      "domainModel": {
        "__delta": {
          "deltas": [
            {
              "key": "aggregates",
              "type": 4,
            },
          ],
          "op": 4,
        },
        "aggregates": [
          {
            "__delta": {
              "deltas": [
                {
                  "key": "root",
                  "type": 4,
                },
                {
                  "key": "entities",
                  "type": 4,
                },
              ],
              "op": 4,
            },
            "commands": [
              {
                "category": "create",
                "description": "",
                "event": {
                  "name": "UserCreateEvent",
                  "properties": [
                    {
                      "access": "public",
                      "description": "",
                      "meta": {},
                      "name": "id",
                      "title": "用户 ID",
                      "type": "Long",
                      "uuid": "1537b249-0aba-4177-8b76-93b1b4a7b535",
                    },
                  ],
                  "uuid": "c89fb0b5-7604-4863-a754-af5a36d6e80f",
                },
                "eventSendable": false,
                "meta": {},
                "name": "UserCreate",
                "properties": [
                  {
                    "access": "public",
                    "description": "",
                    "meta": {},
                    "name": "name",
                    "title": "名称",
                    "type": "String",
                    "uuid": "1a925506-00f7-4aef-bcce-836215911c00",
                  },
                ],
                "repository": "create",
                "return": {
                  "type": "void",
                },
                "rules": [],
                "source": [
                  {
                    "type": "http",
                  },
                  {
                    "type": "rpc",
                  },
                ],
                "title": "新增用户",
                "uuid": "c89fb0b5-7604-4863-a754-af5a36d6e80f",
              },
              {
                "category": "modify",
                "description": "",
                "event": {
                  "name": "UserModifyEvent",
                  "properties": [],
                  "uuid": "e07d602b-b97a-492c-a46e-8798ff3bf8d0",
                },
                "eventSendable": false,
                "meta": {},
                "name": "UserModify",
                "properties": [
                  {
                    "access": "public",
                    "description": "",
                    "meta": {},
                    "name": "id",
                    "title": "用户 ID",
                    "type": "Long",
                    "uuid": "30c6d04a-8ea0-466f-ab4e-d55d9e59d5d7",
                  },
                  {
                    "access": "public",
                    "description": "",
                    "meta": {},
                    "name": "name",
                    "title": "名称",
                    "type": "String",
                    "uuid": "39881d97-37e2-4f33-9220-24827e585aee",
                  },
                ],
                "repository": "modify",
                "return": {
                  "type": "void",
                },
                "rules": [],
                "source": [
                  {
                    "type": "http",
                  },
                  {
                    "type": "rpc",
                  },
                ],
                "title": "编辑用户",
                "uuid": "e07d602b-b97a-492c-a46e-8798ff3bf8d0",
              },
              {
                "category": "remove",
                "description": "",
                "event": {
                  "name": "UserRemoveEvent",
                  "properties": [],
                  "uuid": "a69ea6f4-cc38-4597-b47e-51bf8081f5ea",
                },
                "eventSendable": false,
                "meta": {},
                "name": "UserRemove",
                "properties": [
                  {
                    "access": "public",
                    "description": "",
                    "meta": {},
                    "name": "id",
                    "title": "用户 ID",
                    "type": "Long",
                    "uuid": "25106aaf-6596-44bd-81f5-053a33f5d440",
                  },
                ],
                "repository": "remove",
                "return": {
                  "type": "void",
                },
                "rules": [],
                "source": [
                  {
                    "type": "http",
                  },
                  {
                    "type": "rpc",
                  },
                ],
                "title": "删除用户",
                "uuid": "a69ea6f4-cc38-4597-b47e-51bf8081f5ea",
              },
            ],
            "entities": [
              {
                "__delta": {
                  "deltas": [],
                  "op": 1,
                },
                "id": {
                  "access": "public",
                  "description": "",
                  "meta": {},
                  "name": "id",
                  "title": "用户 ID",
                  "type": "Long",
                  "uuid": "d5ac8416-5fe0-4407-b7d9-c22f207ab3bb",
                },
                "meta": {},
                "methods": [],
                "name": "Order",
                "properties": [
                  {
                    "access": "public",
                    "description": "",
                    "meta": {},
                    "name": "orderId",
                    "title": "订单名称",
                    "type": "String",
                    "uuid": "a59edff9-b80b-43a2-b65d-f1a359acee6c",
                  },
                  {
                    "access": "public",
                    "description": "",
                    "meta": {},
                    "name": "userId",
                    "title": "关联用户",
                    "type": "Long",
                    "uuid": "6159b563-a6f2-478f-864a-828bca126793",
                  },
                ],
                "title": "订单",
                "uuid": "copy-407908de-4cdc-4076-9ec7-958295d9b3e8",
              },
            ],
            "enums": [],
            "meta": {},
            "name": "User",
            "root": {
              "__delta": {
                "deltas": [
                  {
                    "key": "properties",
                    "type": 4,
                  },
                ],
                "op": 4,
              },
              "id": {
                "access": "public",
                "description": "",
                "meta": {},
                "name": "id",
                "title": "用户 ID",
                "type": "Long",
                "uuid": "d5ac8416-5fe0-4407-b7d9-c22f207ab3bb",
              },
              "meta": {},
              "methods": [],
              "name": "User",
              "properties": [
                {
                  "access": "public",
                  "description": "",
                  "meta": {},
                  "name": "name",
                  "title": "名称",
                  "type": "String",
                  "uuid": "9ee1c90d-ac17-45be-897a-5757d1b5f9d7",
                },
                {
                  "__delta": {
                    "deltas": [],
                    "op": 1,
                  },
                  "access": "public",
                  "description": "",
                  "meta": {},
                  "name": "age",
                  "title": "未命名",
                  "type": "Integer",
                  "uuid": "5741d468-0056-4850-a11c-ba9d9c03b8e6",
                },
              ],
              "title": "用户",
              "uuid": "6c75e505-03d9-40aa-9961-bc5e37473da1",
            },
            "title": "用户",
            "uuid": "51d6a0f1-cfef-4723-9aae-2bed0ecf0e9d",
            "valueObjects": [],
          },
        ],
      },
      "name": "simple",
      "objectMapper": {
        "mappers": [],
      },
      "queryModel": {
        "dtos": [],
        "queries": [],
      },
      "ubiquitousLanguage": [
        {
          "conception": "资源",
          "definition": "BD进行销售活动的对象",
          "englishName": "Resource",
          "example": "",
          "restraint": "",
          "uuid": "bf0b0ce7-6028-41d8-89d6-d096485f3f7d",
        },
        {
          "conception": "公海",
          "definition": "业务目标限定的模拟资源全集",
          "englishName": "PublicSea",
          "example": "业务目标是团购买单合作，限定POI为到餐全国和综合58同城的",
          "restraint": "与业务目标对应，比如交易公海是为了做交易，推广公海是为了推广",
          "uuid": "c9bae1d0-d028-418e-98f6-6a3990eec09a",
        },
      ],
      "uuid": "simple",
      "version": "0.1.0",
      "vision": "",
    },
  ],
  "businessScenarios": [],
  "name": "应用",
  "uuid": "application",
  "version": "1.0.0",
}
`;
